bokehjs/src/lib/models/glyphs/webgl/base_marker.ts

```diff
@@ -0,0 +1,152 @@
+
+  // visual properties, either all or none are set.
+  private _linewidths?: Float32Buffer
+  private _line_rgba: NormalizedUint8Buffer
+  private _fill_rgba: NormalizedUint8Buffer
+  private _line_caps: Uint8Buffer
+  private _line_joins: Uint8Buffer
+
+  // Only needed if have hatch pattern, either all or none of the buffers are set.
+  private _have_hatch: boolean
+  private _hatch_patterns?: Uint8Buffer
+  private _hatch_scales?: Float32Buffer
+  private _hatch_weights?: Float32Buffer
+  private _hatch_rgba?: NormalizedUint8Buffer
+
+  // Avoiding use of nan or inf to represent missing data in webgl as shaders may
+  // have reduced floating point precision.  So here using a large-ish negative
+  // value instead.
+  protected static readonly missing_point = -10000
+
+  constructor(regl_wrapper: ReglWrapper, override readonly glyph: GlyphView) {
+    super(regl_wrapper, glyph)
+
+    this._antialias = 1.5
+    this._show_all = false
+  }
+
+  abstract override draw(indices: number[], mainglyph: GlyphView, trans: Transform): void
+
+  protected _draw_one_marker_type(marker_type: MarkerType, transform: Transform, main_gl_glyph: BaseMarkerGL): void {
+    if (this._have_hatch) {
+      const props: MarkerHatchGlyphProps = {
+        scissor: this.regl_wrapper.scissor,
+        viewport: this.regl_wrapper.viewport,
+        canvas_size: [transform.width, transform.height],
+        pixel_ratio: transform.pixel_ratio,
+        center: main_gl_glyph._centers!,
+        width: main_gl_glyph._widths!,
+        height: main_gl_glyph._heights!,
+        angle: main_gl_glyph._angles!,
+        size_hint: marker_type_to_size_hint(marker_type),
+        nmarkers: main_gl_glyph.nvertices,
+        antialias: this._antialias,
+        linewidth: this._linewidths!,
+        line_color: this._line_rgba,
+        fill_color: this._fill_rgba,
+        line_cap: this._line_caps,
+        line_join: this._line_joins,
+        show: this._show!,
+        hatch_pattern: this._hatch_patterns!,
+        hatch_scale: this._hatch_scales!,
+        hatch_weight: this._hatch_weights!,
+        hatch_color: this._hatch_rgba!,
+      }
+      this.regl_wrapper.marker_hatch(marker_type)(props)
+    } else {
+      const props: MarkerGlyphProps = {
+        scissor: this.regl_wrapper.scissor,
+        viewport: this.regl_wrapper.viewport,
+        canvas_size: [transform.width, transform.height],
+        pixel_ratio: transform.pixel_ratio,
+        center: main_gl_glyph._centers!,
+        width: main_gl_glyph._widths!,
+        height: main_gl_glyph._heights!,
+        angle: main_gl_glyph._angles!,
+        size_hint: marker_type_to_size_hint(marker_type),
+        nmarkers: main_gl_glyph.nvertices,
+        antialias: this._antialias,
+        linewidth: this._linewidths!,
+        line_color: this._line_rgba,
+        fill_color: this._fill_rgba,
+        line_cap: this._line_caps,
+        line_join: this._line_joins,
+        show: this._show!,
```

Comment:
Is there not ant way in TS to effectively factor out the overlapping structure assignments so that they are not duplicated?

Reply:
Yes there is, new commit is on its way.

This can be applied in a number of places. It could make a big improvement in `regl_wrap.ts`, although it will be a bit trickier there. That is probably a separate PR.